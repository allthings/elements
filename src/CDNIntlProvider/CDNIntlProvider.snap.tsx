// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[
  `Check the CDNIntlProvider component should fetch the corresponding locales 1`
] = `
Array [
  <span
    data-state="{\\"test\\":\\"Hallo Welt\\"}"
    id="__ELEMENTS_INTL__"
    style={
      Object {
        "display": "none",
      }
    }
  />,
  <span>
    Hallo Welt
  </span>,
]
`

exports[
  `Check the CDNIntlProvider component should fetch the corresponding locales 2`
] = `
Array [
  <span
    data-state="{\\"test\\":\\"Hello World\\"}"
    id="__ELEMENTS_INTL__"
    style={
      Object {
        "display": "none",
      }
    }
  />,
  <span>
    Hello World
  </span>,
]
`

exports[
  `Check the CDNIntlProvider component should pick up if hyrdated 1`
] = `"<span id=\\"__ELEMENTS_INTL__\\" style=\\"display:none\\" data-state=\\"{&quot;test&quot;:&quot;Bonjour tout le monde et ala&quot;}\\"></span><span>Bonjour tout le monde et ala</span>"`

exports[
  `Check the CDNIntlProvider component should use the provided locale and do no fetch 1`
] = `
Array [
  <span
    data-state="{\\"test\\":\\"Bonjour tout le monde\\"}"
    id="__ELEMENTS_INTL__"
    style={
      Object {
        "display": "none",
      }
    }
  />,
  <span>
    Bonjour tout le monde
  </span>,
]
`
