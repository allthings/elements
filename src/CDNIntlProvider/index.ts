export { default } from './CDNIntlProvider'
export * from './CDNIntlProvider'
