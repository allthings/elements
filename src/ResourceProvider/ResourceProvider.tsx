import * as React from 'react'

export const RESOURCE_PATH = 'https://static.allthings.me'

/**
 * Elements uses a set of static resources like images or icons.
 * In order to benefit from caching across all apps, these resources are provided by a static asset CDN.
 *
 * Whenever you like to use Icons or Illustrations, you need to use the ResourceProvider to let the components know where they are.
 */
export const ResourceProviderContext = React.createContext({
  resourcePath: RESOURCE_PATH,
})

export interface IResourceProviderProps {
  readonly resourcePath?: string
}

/**
 * All elements support theming by default, and therefore every element must be wrapped inside a ThemeProvider.
 * The ThemeProvider allows you to define the default colors for most elements.
 *
 * **Example**: If you want all you buttons to be red, instead of writing <Button color="red" /> all the time, you might want to set the "primary" color of your theme to red.
 **/
const ResourceProvider: React.FC<IResourceProviderProps> = ({
  children,
  resourcePath = RESOURCE_PATH,
}) => {
  // this is to prevent unintentional re-renders in consumers
  // see https://reactjs.org/docs/context.html#caveats
  const [value, setValue] = React.useState({ resourcePath })
  React.useEffect(() => {
    setValue({ resourcePath })
  }, [resourcePath])

  return (
    <ResourceProviderContext.Provider value={value}>
      {children}
    </ResourceProviderContext.Provider>
  )
}

export default ResourceProvider
