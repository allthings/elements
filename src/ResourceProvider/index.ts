export { default } from './ResourceProvider'
export * from './ResourceProvider'
