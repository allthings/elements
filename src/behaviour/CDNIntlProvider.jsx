import PropTypes from 'prop-types'
import React, { Fragment } from 'react'
import { IntlProvider } from 'react-intl'
import 'cross-fetch/polyfill'

export const loadLanguage = async (
  resourcePath,
  project,
  variation,
  locale,
  stage
) => {
  const countryCode = locale.split('_')[0]
  const translations = await fetch(
    `${resourcePath}/${project}/${stage}/i18n/${countryCode}/${variation}.json`
  )
  return translations.json()
}

class CDNIntlProvider extends React.Component {
  static propTypes = {
    children: PropTypes.node,
    /** Locale you like to get, EN_us, DE_de */
    locale: PropTypes.string.isRequired,
    /** Optionally pass messages. This will prevent initial loading. */
    messages: PropTypes.object,
    /** Called when new languages got loaded */
    onDone: PropTypes.func,
    /** The project ID loading the langauges for */
    project: PropTypes.string.isRequired,
    /** Stage, can be production and staging */
    stage: PropTypes.oneOf(['prerelease', 'production', 'staging']),
    /** "Default" by default. Can be any allowed variation string.  */
    variation: PropTypes.string,
  }

  static defaultProps = {
    onDone: _ => _,
    stage: 'production',
    variation: 'default',
  }

  static contextTypes = {
    resourcePath: PropTypes.string,
  }

  constructor(props, context) {
    super(props, context)
    const messages = props.messages
    this.state = {
      messages,
    }
    if (!messages && typeof document !== 'undefined') {
      const container = document.getElementById('__ELEMENTS_INTL__')
      if (container) {
        const stateString = container.getAttribute('data-state')
        this.state = { messages: JSON.parse(stateString) }
      } else {
        this.loadLanguages(props)
      }
    }
  }

  componentWillReceiveProps(nextProps) {
    const { project, locale, variation } = this.props
    if (
      nextProps.project !== project ||
      nextProps.locale !== locale ||
      nextProps.variation !== variation
    ) {
      this.loadLanguages(nextProps)
    }
  }

  loadLanguages = async props => {
    const { project, variation, locale, stage, onDone } = props

    const messages = await loadLanguage(
      this.context.resourcePath,
      project,
      variation,
      locale,
      stage
    )

    this.setState({ messages }, onDone)
  }

  renderSideEffect = messages => (
    <span
      id="__ELEMENTS_INTL__"
      style={{ display: 'none' }}
      data-state={messages && JSON.stringify(messages)}
    />
  )

  render() {
    const countryCode = this.props.locale.split('_')[0]
    const { messages } = this.state

    return (
      <IntlProvider locale={countryCode} messages={messages}>
        <Fragment>
          {this.renderSideEffect(messages)}
          {messages && this.props.children}
        </Fragment>
      </IntlProvider>
    )
  }
}

export default CDNIntlProvider
