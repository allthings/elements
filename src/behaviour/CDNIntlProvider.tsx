import PropTypes from 'prop-types'
import React, { Fragment, Component } from 'react'
import { IntlProvider } from 'react-intl'

export const loadLanguage = async (
  resourcePath: string,
  project: string,
  variation: string,
  locale: string,
  stage: string
): Promise<object> => {
  const countryCode = locale.split('_')[0]
  const translations = await fetch(
    `${resourcePath}/${project}/${stage}/i18n/${countryCode}/${variation}.json`
  )
  return translations.json()
}

interface IBasicProps {
  project: string
  locale: string
  onDone?: () => void
  messages?: object
  stage: 'prerelease' | 'production' | 'staging'
  variation?: string
}

const defaultProps = {
  onDone: () => {},
  stage: 'production',
  variation: 'default',
}

type IProps = IBasicProps & typeof defaultProps

interface IState {
  messages?: object
}

class CDNIntlProvider extends Component<IProps, IState> {
  static defaultProps = defaultProps

  static contextTypes = {
    resourcePath: PropTypes.string,
  }

  constructor(props: IProps, context: any) {
    super(props, context)
    const messages = props.messages
    this.state = {
      messages,
    }
    if (!messages && typeof document !== 'undefined') {
      const container = document.getElementById('__ELEMENTS_INTL__')
      if (container) {
        const stateString = container.getAttribute('data-state') as string
        this.state = { messages: JSON.parse(stateString) }
      } else {
        this.loadLanguages(props)
      }
    }
  }

  componentDidUpdate(prevProps: IProps) {
    const { project, locale, variation } = this.props
    if (
      prevProps.project !== project ||
      prevProps.locale !== locale ||
      prevProps.variation !== variation
    ) {
      this.loadLanguages(this.props)
    }
  }

  loadLanguages = async (props: IProps) => {
    const { project, variation, locale, stage, onDone } = props

    const messages = await loadLanguage(
      this.context.resourcePath,
      project,
      variation,
      locale,
      stage
    )

    this.setState({ messages }, onDone)
  }

  renderSideEffect = (messages?: object) => (
    <span
      id="__ELEMENTS_INTL__"
      style={{ display: 'none' }}
      data-state={messages && JSON.stringify(messages)}
    />
  )

  render() {
    const countryCode = this.props.locale.split('_')[0]
    const { messages } = this.state

    return (
      <IntlProvider locale={countryCode} messages={messages}>
        <Fragment>
          {this.renderSideEffect(messages)}
          {messages && this.props.children}
        </Fragment>
      </IntlProvider>
    )
  }
}

export default CDNIntlProvider
