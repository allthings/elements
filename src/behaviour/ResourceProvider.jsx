import React from 'react'
import PropTypes from 'prop-types'

export const RESOURCE_PATH = 'https://static.allthings.me'

/**
 * Elements uses a set of static resources like images or icons.
 * In order to benefit from caching across all apps, these resources are provided by a static asset CDN.
 *
 * Whenever you like to use Icons or Illustrations, you need to use the ResouceProvider to let the components know where they are.
 */
class ResourceProvider extends React.Component {
  getChildContext() {
    return { resourcePath: this.props.resourcePath }
  }

  static childContextTypes = {
    resourcePath: PropTypes.string,
  }

  static propTypes = {
    resourcePath: PropTypes.string.isRequired,
    children: PropTypes.node.isRequired,
  }

  static defaultProps = {
    resourcePath: RESOURCE_PATH,
    children: null,
  }

  render() {
    return this.props.children
  }
}

export default ResourceProvider
