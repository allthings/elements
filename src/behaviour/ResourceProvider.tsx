import React from 'react'

export const RESOURCE_PATH = 'https://static.allthings.me'

/**
 * Elements uses a set of static resources like images or icons.
 * In order to benefit from caching across all apps, these resources are provided by a static asset CDN.
 *
 * Whenever you like to use Icons or Illustrations, you need to use the ResouceProvider to let the components know where they are.
 */
const ResouceProvider = React.createContext({
  resourcePath: RESOURCE_PATH,
})

export type ResourceType = {
  resourcePath?: string
}

export default ResouceProvider
