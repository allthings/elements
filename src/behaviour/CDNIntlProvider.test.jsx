import React from 'react'
import renderer from 'react-test-renderer'
import { FormattedMessage } from 'react-intl'
import CDNIntlProvider from './CDNIntlProvider'
import fetch from 'jest-fetch-mock'
import ResourceProvider from './ResourceProvider'

describe('Check the CDNIntlProvider component', () => {
  it('should fetch the corresponding locales', async () => {
    const fetchGerman = fetch.mockResponse(
      JSON.stringify({ test: 'Hallo Welt' })
    )

    const testRenderer = await new Promise((resolve, reject) => {
      let myRenderer
      function onDone() {
        resolve(myRenderer)
      }
      const nbm = (
        <ResourceProvider>
          <CDNIntlProvider
            locale="de_DE"
            project="app"
            variation="residential-formal"
            onDone={onDone}
          >
            <FormattedMessage id="test" defaultMessage="Default" />
          </CDNIntlProvider>
        </ResourceProvider>
      )

      myRenderer = renderer.create(nbm)
    })

    expect(fetchGerman).toHaveBeenCalledWith(
      'https://static.allthings.me/app/staging/i18n/de/residential-formal.json'
    )
    expect(testRenderer).toMatchSnapshot()

    const fetchEnglish = fetch.mockResponse(
      JSON.stringify({ test: 'Hello World' })
    )

    await new Promise((resolve, reject) => {
      const nbm = (
        <ResourceProvider>
          <CDNIntlProvider
            locale="en_US"
            project="app"
            variation="residential-formal"
            onDone={resolve}
          >
            <FormattedMessage id="test" defaultMessage="Default" />
          </CDNIntlProvider>
        </ResourceProvider>
      )

      testRenderer.update(nbm)
    })

    expect(fetchEnglish).toHaveBeenCalledWith(
      'https://static.allthings.me/app/staging/i18n/en/residential-formal.json'
    )
    expect(testRenderer).toMatchSnapshot()
  })
  it('should use the provided locale and do no fetch', () => {
    const onDone = jest.fn()
    const wrapper = mount(
      <ResourceProvider>
        <CDNIntlProvider
          messages={{ test: 'Bonjour tout le monde' }}
          locale="fr_FR"
          project="app"
          variation="residential-formal"
          onDone={onDone}
        >
          <FormattedMessage id="test" defaultMessage="Default" />
        </CDNIntlProvider>
      </ResourceProvider>
    )
    expect(wrapper).toMatchSnapshot()
    expect(onDone.mock.calls.length).toBe(0)
  })
})
