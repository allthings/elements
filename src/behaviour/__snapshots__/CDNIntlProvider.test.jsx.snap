// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check the CDNIntlProvider component should fetch the corresponding locales 1`] = `
<span>
  Hallo Welt
</span>
`;

exports[`Check the CDNIntlProvider component should fetch the corresponding locales 2`] = `
<span>
  Hello World
</span>
`;

exports[`Check the CDNIntlProvider component should use an alternative fetch method 1`] = `
<span>
  こんにちは世界
</span>
`;

exports[`Check the CDNIntlProvider component should use the correct stage 1`] = `
<CDNIntlProvider
  fetchMethod={[MockFunction]}
  locale="fr_FR"
  messages={
    Object {
      "test": "Bonjour tout le monde",
    }
  }
  onDone={[Function]}
  project="app"
  stage="staging"
  variation="residential-formal"
>
  <FormattedMessage
    defaultMessage="Default"
    id="test"
    values={Object {}}
  />
</CDNIntlProvider>
`;

exports[`Check the CDNIntlProvider component should use the provided locale and do no fetch 1`] = `
<span>
  Bonjour tout le monde
</span>
`;
